grpc/src/builder.rs

```diff
@@ -0,0 +1,188 @@
+use bytes::Bytes;
+use http_body::Body;
+use k256::ecdsa::VerifyingKey;
+use signature::Keypair;
+use tonic::body::BoxBody;
+use tonic::client::GrpcService;
+#[cfg(not(target_arch = "wasm32"))]
+use tonic::transport::{Channel, ClientTlsConfig, Endpoint};
+
+use crate::grpc::StdError;
+use crate::{DocSigner, GrpcClient, Result, TxClient};
+
+#[cfg(not(target_arch = "wasm32"))]
+#[derive(thiserror::Error, Debug)]
+pub enum GrpcClientBuilderCertError {
+    /// Error handling certificate root
+    #[error(transparent)]
+    WebpkiError(#[from] webpki::Error),
+
+    /// Could not import system certificates
+    #[error("Could not import platform certificates: {errors:?}")]
+    RustlsNativeCertsError {
+        errors: Vec<rustls_native_certs::Error>,
+    },
+}
+
+/// Builder for [`GrpcClient`] and [`TxClient`]
+#[derive(Clone)]
+pub struct GrpcClientBuilder<T, S> {
+    connection: T,
+    signer: S,
+    account_pubkey: Option<VerifyingKey>,
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl GrpcClientBuilder<(Endpoint, ClientTlsConfig), ()> {
+    /// Create a new client connected to the given `url` using [`Channel`] transport
+    ///
+    /// [`Channel`]: tonic::transport::Channel
+    pub fn with_url(url: impl Into<String>) -> Result<Self, tonic::transport::Error> {
+        let endpoint = Endpoint::from_shared(url.into())?.user_agent("celestia-grpc")?;
+        let tls_config = ClientTlsConfig::new();
```

Comment:
this applies to almost all methods, can we just store information we need and move all the fallible logic to happen at the final call to build?